### PR TITLE
test: v1 testing infrastructure — coverage gaps, enforcement, workflow

### LINKED ISSUE #130: test: add unit tests for Migration import phase
## Summary

Add unit tests for the Migration import phase. This is critical path code with **0 unit test coverage**.

**Priority:** P1 (critical path, orchestrates entire import process)

## Background

The import phase handles:
- Tiered import orchestration (dependency ordering)
- Plugin step suspension/restoration during import
- Referential integrity via relationship processing
- Schema validation before import
- Deferred field handling for circular references

This is complex business logic that needs unit test coverage with mocked dependencies.

### Import Phase (`src/PPDS.Migration/Import/`)
- [ ] `TieredImporter.cs` - orchestrates entire import process
- [ ] `PluginStepManager.cs` - plugin suspension/restoration
- [ ] `RelationshipProcessor.cs` - referential integrity handling
- [ ] `SchemaValidator.cs` - schema validation logic
- [ ] `DeferredFieldProcessor.cs` - deferred field handling
- [ ] `ImportTier1.cs` - tier 1 import logic
- [ ] `ImportTier2.cs` - tier 2 import logic
- [ ] `ImportTier3.cs` - tier 3 import logic

### Schema (`src/PPDS.Migration/Schema/`)
- [ ] `DataverseSchemaGenerator.cs` - schema generation

### User Mapping (`src/PPDS.Migration/UserMapping/`)
- [ ] `UserMappingGenerator.cs` - user mapping generation

## Test Approach

- Create `tests/PPDS.Migration.Tests/Import/` directory structure
- Mock `IDataverseConnectionPool` and `IOrganizationService`
- Test each class in isolation
- Cover success paths, error handling, and edge cases

## Acceptance Criteria

- [ ] Each file has corresponding test class
- [ ] Tests use mocked dependencies (no live Dataverse)
- [ ] Critical paths have multiple test cases
- [ ] Error handling is tested
- [ ] All tests pass in CI

## Related

- Parent epic: `epic:testing`
- Related: #104 (CLI E2E tests)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### LINKED ISSUE #658: fix(test): PluginsDiffCommandE2ETests.Diff_AfterDeploy_ShowsNoDrift flaky on CI
## Summary

`PluginsDiffCommandE2ETests.Diff_AfterDeploy_ShowsNoDrift` is flaky in CI. It depends on live Dataverse environment state — plugin step registration — which is non-deterministic across runs.

## Evidence

- **#652** (migration polish): FAIL — same test, no plugin code changes
- **#654** (v1 bugs fix): PASS — no plugin code changes
- **#655** (webresources CLI): FAIL — zero plugin files touched

The test deploys test plugins, then runs `ppds plugins diff` expecting no drift. The "Missing step" error indicates the deploy didn't fully register before the diff check ran.

## Root Cause

Server-side timing — the deploy step registers plugins asynchronously on the Dataverse platform, and the diff check runs before registration completes. This is environment state, not a code regression.

## Suggested Fix

Either add a retry/wait loop after deploy before running diff, or make the test assert on the deploy output first to confirm registration succeeded before checking drift.

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### LINKED ISSUE #670: infra: Enforce Codecov patch coverage as PR gate
## Description

Configure Codecov to enforce patch (new code) coverage as a required PR status check, preventing coverage erosion on new contributions.

## Problem

Coverage is currently informational only — component targets exist (Plugins 95%, Auth 70%, Migration 50%, etc.) but PRs can merge with zero test coverage on new code. The testing infrastructure review (2026-03-24) identified this as a process gap: the AI workflow (/qa, /verify) catches current-state issues per-PR, but nothing prevents long-term coverage erosion.

## Solution

1. Change `codecov.yml` patch coverage from informational to required:
   - Target: 80% patch coverage (already configured, just not enforced)
   - Set `after_n_builds` to match the number of CI upload jobs

2. Add Codecov as a required status check in GitHub branch protection for `main`

3. Document the coverage policy in CLAUDE.md or testing docs

## Acceptance Criteria

- [ ] PRs with <80% coverage on new code show a failing Codecov status check
- [ ] Branch protection requires Codecov patch check to pass
- [ ] Existing low-coverage areas are not penalized (only new/changed code)
- [ ] Coverage policy is documented

## Notes

- Component-level thresholds can remain informational for now
- Consider making the threshold 70% initially to avoid friction, then ratchet to 80%
- Analyzers (#447 PPDS007, #448 PPDS008) complement this with static guarantees

### LINKED ISSUE #673: infra: Mandate test writing in /implement and AC verification in /gates
## Description

Patch the AI workflow to close the test-writing gap: /implement should produce tests for every spec AC, and /gates should mechanically verify the mapping.

## Problem

The spec template has an AC table with a `Test` column (AC-01 → TestClass.TestMethod), but nothing in the pipeline enforces that test methods get created during implementation. Tests are optional in practice, which means features can ship without regression coverage.

### 1. Patch /implement skill

Add a required step after implementation: for each AC in the spec, produce a corresponding test method. Update the spec's AC table with the test mapping and status.

Specifically:
- Read the spec's AC table
- For each AC, write a test that validates the criterion
- Update the spec's AC table: fill in the `Test` column with `TestClass.TestMethod` and set status to ✅
- Tests must use appropriate category traits (`Unit`, `TuiUnit`, `Integration`, etc.)

### 2. Patch /gates skill

Add an AC coverage check after existing gates (build, lint, test):
- Parse the spec's AC table
- For each AC with a `Test` column entry, verify the test method exists
- For each AC with a `Test` column entry, verify the test passes
- Report: "AC coverage: N/M criteria have passing tests"
- Fail the gate if any mapped test doesn't exist or doesn't pass

### Design Decision: Why hybrid, not a separate skill?

- The implementer has the deepest context — tests written alongside code are higher quality
- A separate /test-write skill would re-derive all the same context (slower, shallower)
- Implementation bias is mitigated by /impartial-code-review catching tests that don't validate the AC
- /gates provides the mechanical safety net
- A standalone test-writing skill can be reconsidered if backfill needs grow

## Acceptance Criteria

- [ ] /implement skill includes test-writing step for each spec AC
- [ ] /implement updates the spec AC table with test mappings
- [ ] /gates verifies AC → test method existence and pass status
- [ ] /gates reports AC coverage ratio
- [ ] Existing specs with no tests don't block /gates (only checks mapped tests)

## Notes

- Part of testing infrastructure review (2026-03-24)
- Does not require a new skill — modifies two existing skills
- For backfilling tests on existing code (#130, etc.), the implementer handles test writing as part of the issue work

### LINKED ISSUE #674: infra: Add .NET unit tests to pre-commit hook
## Description

Extend the pre-commit hook to run .NET unit tests on changed projects, closing the gap where .NET-breaking changes pass the hook but fail in CI.

## Problem

The pre-commit hook (`scripts/hooks/pre-commit`) currently only gates TypeScript:
- Typecheck (`npm run typecheck:all`)
- ESLint (`npx eslint src --quiet`)

.NET changes pass through with zero local validation. An AI agent running /implement can commit code that breaks .NET tests without knowing until CI reports back — a feedback loop of minutes instead of seconds.

## Solution

When staged files include `.cs` or `.csproj` changes, run:

```bash
dotnet test PPDS.sln --filter "Category!=Integration&Category!=SecureStorage&Category!=SlowIntegration&Category!=DestructiveE2E" -v q --no-build
```

### Design Considerations

1. **Speed**: The full unit suite runs in ~2-3 seconds (no-build). With build, ~15-20 seconds. This is acceptable for a pre-commit gate.

2. **Scope**: Run all unit tests, not just tests for changed projects. The test suite is fast enough that scoping adds complexity without meaningful time savings.

3. **`--no-build`**: Use `--no-build` to avoid double-building. The hook should assume the developer (or AI agent) has already built. If the build is stale, CI catches it.

4. **Skip mechanism**: Respect `--no-verify` for cases where developers need to commit work-in-progress. This is already standard git behavior.

## Acceptance Criteria

- [ ] Pre-commit hook runs .NET unit tests when `.cs` or `.csproj` files are staged
- [ ] Hook skips .NET tests when no .NET files are staged
- [ ] Hook blocks commit if any .NET unit test fails
- [ ] Hook displays clear error message on failure
- [ ] TypeScript gates continue to work unchanged
- [ ] Total hook runtime stays under 30 seconds

## Notes

- Part of testing infrastructure review (2026-03-24)
- Consider whether `dotnet build` should also be part of the hook (catches compile errors before test run)
- The existing hook already handles the "no staged files" skip pattern for TypeScript — follow the same pattern for .NET

### PR DESCRIPTION
**Pull request by joshsmithxrm**
## Summary

- **Testing standards enforcement:** Constitution I6 (AC→test mandate), /implement AC gate hardened, pre-commit .NET build+test gate, Codecov 80% patch coverage enforced
- **Coverage gap closure:** 50 files, ~7,000 lines — import orchestration (TieredImporter, RelationshipProcessor, SchemaValidator, etc.), plugin registration (unregister/delete/get/list), SqlLanguageService, 4 Extension panels, 28 MCP tools
- **Infrastructure:** main-branch protection hook (blocks Edit/Write on main), /design skill plan-commit fix, pipeline timeout fix

Closes #130
Closes #658
Closes #670
Closes #673
Closes #674

## Test Plan

- [x] All .NET unit tests pass (6,792 across net8.0/9.0/10.0)
- [x] All Extension Vitest tests pass (409 across 25 files)
- [x] Pre-commit hook runs .NET build + tests on staged .cs files
- [x] Pre-commit hook runs TS typecheck + lint on staged .ts files
- [x] Codecov patch coverage set to 80% enforced
- [x] Constitution I6 added and referenced in /implement AC gate
- [x] protect-main-branch.py blocks Edit/Write on main, allows .plans/ and temp
- [x] Flaky PluginsDiff E2E test fixed with polling loop

## Verification

- [x] /gates passed
- [x] /verify completed (surfaces: infra, ext, cli, mcp)
- [x] /qa completed (surfaces: infra, cli, ext, mcp)
- [x] /review completed (findings: 8, all resolved in converge)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .claude/hooks/protect-main-branch.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        result = subprocess.run(
+            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
+            cwd=os.environ.get("CLAUDE_PROJECT_DIR", "."),
+            capture_output=True,
+            text=True,
+            timeout=5,
+        )
+        return result.stdout.strip()
+    except Exception:
+        return ""
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply .claude/hooks/protect-main-branch.py:28]: Fixed in 9d1786852 — replaced bare `except Exception` with specific `(subprocess.TimeoutExpired, FileNotFoundError)` catches, added `returncode != 0` check with stderr warning. Fail-open behavior preserved but failures are now visible.

File: tests/PPDS.Migration.Tests/Import/DeferredFieldProcessorTests.cs [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+                "account",
+                It.IsAny<IEnumerable<Entity>>(),
+                It.IsAny<BulkOperationOptions>(),
+                It.IsAny<IProgress<ProgressSnapshot>>(),
+                It.IsAny<CancellationToken>()))
+            .ReturnsAsync((string _, IEnumerable<Entity> entities, BulkOperationOptions _, IProgress<ProgressSnapshot> _, CancellationToken _) =>
+            {
+                var list = entities.ToList();
+                // Verify the entity has the remapped reference
+                var updateEntity = list[0];
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply tests/PPDS.Migration.Tests/Import/DeferredFieldProcessorTests.cs]: Fixed in ebc0afb96 — removed unused variable.

File: tests/PPDS.Migration.Tests/Import/TieredImporterTests.cs [hunk 3]
```diff
@@ -0,0 +1,1373 @@
...
+        var result = await _sut.ImportAsync(data, plan, options);
+
+        // Assert
+        result.RecordsImported.Should().Be(2);
+        // The contact's parentcustomerid should still reference the account
+        // (in this case IDs are preserved since TieredImporter maps oldId -> oldId)
+        capturedContact.Should().NotBeNull();
+        capturedContact!.Contains("parentcustomerid").Should().BeTrue();
+        var lookup = capturedContact["parentcustomerid"] as EntityReference;
+        lookup.Should().NotBeNull();
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply tests/PPDS.Migration.Tests/Import/TieredImporterTests.cs]: Fixed in ebc0afb96 — replaced `as EntityReference` + null assertion with throw expression for non-null guarantee.

File: tests/PPDS.Migration.Tests/Import/TieredImporterTests.cs [hunk 4]
```diff
@@ -0,0 +1,1373 @@
...
+
+        // Act
+        var result = await _sut.ImportAsync(data, plan, options);
+
+        // Assert
+        result.RecordsImported.Should().Be(1);
+        capturedEntity.Should().NotBeNull();
+        capturedEntity!.Contains("ownerid").Should().BeTrue();
+        var ownerRef = capturedEntity["ownerid"] as EntityReference;
+        ownerRef.Should().NotBeNull();
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply tests/PPDS.Migration.Tests/Import/TieredImporterTests.cs]: Fixed in ebc0afb96 — same pattern, throw expression instead of null assertion.

File: tests/PPDS.Migration.Tests/Import/TieredImporterTests.cs [hunk 5]
```diff
@@ -0,0 +1,1373 @@
...
+        };
+
+        // Act
+        var result = await _sut.ImportAsync(data, plan, options);
+
+        // Assert
+        result.RecordsImported.Should().Be(1);
+        capturedEntity.Should().NotBeNull();
+        var ownerRef = capturedEntity!["ownerid"] as EntityReference;
+        ownerRef.Should().NotBeNull();
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply tests/PPDS.Migration.Tests/Import/TieredImporterTests.cs]: Fixed in ebc0afb96 — same pattern, throw expression instead of null assertion.
